generator.go

```diff
@@ -147,11 +151,57 @@ func NewGen() *Gen {
 // MAC address being used, you'll need to create a new generator using this
 // function.
 func NewGenWithHWAF(hwaf HWAddrFunc) *Gen {
-	return &Gen{
+	return NewGenWithOptions(WithHWAddrFunc(hwaf))
+}
+
+// NewGenWithOptions returns a new instance of Gen with the options provided.
+// Most people should use NewGen() or NewGenWithHWAF() instead.
+//
+// To customize the generator, you can pass in one or more GenOption functions.
+// For example:
+//
+//	gen := NewGenWithOptions(
+//	    WithHWAddrFunc(myHWAddrFunc),
+//	    WithEpochFunc(myEpochFunc),
+//	    WithRandomReader(myRandomReader),
+//	)
+//
+// NewGenWithOptions(WithHWAddrFunc(myHWAddrFunc)) is equivalent to calling
+// NewGenWithHWAF(myHWAddrFunc)
+// NewGenWithOptions() is equivalent to calling NewGen()
+func NewGenWithOptions(opts ...GenOption) *Gen {
+	gen := &Gen{
 		epochFunc:  time.Now,
-		hwAddrFunc: hwaf,
+		hwAddrFunc: defaultHWAddrFunc,
 		rand:       rand.Reader,
 	}
+
+	for _, opt := range opts {
+		opt(gen)
+	}
+
+	return gen
+}
+
+// WithHWAddrFunc is a GenOption that allows you to provide your own HWAddrFunc function
+func WithHWAddrFunc(hwaf HWAddrFunc) GenOption {
+	return func(gen *Gen) {
+		gen.hwAddrFunc = hwaf
```

Comment:
This needs a nil check.  `WithHWAddrFunc(nil)` will compile but lead to an immediate panic when the generator is used.

Reply:
We do not need to validate each option, instead, we can bypass `NewGenWithOptions` when the `nil` option has been passed. 

```go
func NewGenWithOptions(opts ...GenOption) *Gen {
	gen := &Gen{
		epochFunc:  time.Now,
		hwAddrFunc: defaultHWAddrFunc,
		rand:       rand.Reader,
	}

	for _, opt := range opts {
		if opt != nil {
			opt(gen)
		}
	}

	return gen
}
```

What do you think?

Reply:
In the example I gave the option itself won't be `nil` but it would end up setting the `Gen.hwAddrFunc` field to `nil`.  The `if opt != nil` check will never be false unless someone explicitly passes `GenOption(nil)`.

To be 100% safe we should be checking the option itself in `NewGenWithOptions()` and checking for `nil` **values** in the option functions themselves.

```
func WithHWAddrFunc(hwaf HWAddrFunc) GenOption {
    return func(gen *Gen) {
        if hwaf != nil {
            gen.hwAddrFunc = hwaf
        }
    }
}
```
or
```
func WithHWAddrFunc(hwaf HWAddrFunc) GenOption {
    return func(gen *Gen) {
        if hwaf == nil {
            hwaf = defaultHWAddrFunc
        }
        gen.hwAddrFunc = hwaf
    }
}
```

Reply:
Got it!

Changed